Add inline asm support for amdgpu

Flakebi opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149793)*
<!-- homu-ignore:end -->

Add support for inline assembly for the amdgpu backend (the
amdgcn-amd-amdhsa target).
Add register classes for `vgpr` (vector general purpose register) and
`sgpr` (scalar general purpose register).
The LLVM backend supports two more classes, `reg`, which is either VGPR
or SGPR, up to the compiler to decide. As instructions often rely on a
register being either a VGPR or SGPR for the assembly to be valid, reg
doesn’t seem that useful (I struggled to write correct tests for it), so
I didn’t end up adding it.
The fourth register class is AGPRs, which only exist on some hardware
versions (not the consumer ones) and they have restricted ways to write
and read from them, which makes it hard to write a Rust variable into
them. They could be used inside assembly blocks, but I didn’t add them
as Rust register class.

There are a few change affecting general inline assembly code, that is
`InlineAsmReg::name()` now returns a `Cow` instead of a `&'static str`.
Because amdgpu has many registers, 256 VGPRs plus combinations of 2 or 4
VGPRs, and I didn’t want to list hundreds of static strings, the amdgpu
reg stores the register number(s) and a non-static String is generated
at runtime for the register name.
Similar for register classes and supported_types.

Vectors of 64-bit types are supported by the LLVM backend, but omitted
here to make the code simpler. There is currently no systematic support
in LLVM of which vectors of 64-bit types are supported. Also, they are
likely seldomly unused, vectors of 16- and 32-bit types are important.

Tracking issue: rust-lang/rust#135024

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

eholk commented:
This seems okay to me, but I'd rather someone more familiar with this part of the compiler give the final signoff.

@bors r?

eholk commented:
@bors r? compiler

fee1-dead commented:
@rustbot reroll

Flakebi commented:
@taiki-e, I saw your mention in the asm tracking issue. amdgpu has named registers (>384 of them), so a clobber_abi would be possible. However, there’s currently no stable ABI, so we would need to follow what the LLVM backend does and I think there’s no tools to make sure the Rust and LLVM ABI understanding is in sync, so that would be asking for subtle breakages. Therefore I think it doesn’t make sense to support clobber_abi with amdgpu at the moment.

Flakebi commented on compiler/rustc_target/src/asm/amdgpu.rs:
```diff
@@ -0,0 +1,234 @@
+def_reg_class! {
+    Amdgpu AmdgpuInlineAsmRegClass {
+        sgpr,
+        vgpr,
```
Hm, they can be used as SIMD/vector registers as well (I added support in #149994), up to `[32 x i32]`.
So, the types would be something like vgpr{16,32,64,96,128,160,192,224,256,288,320,352,384,416,448,480,512,544,576,608,640,672,704,736,768,800,832,864,896,928,960,992,1024}. (a few might not be valid, but the LLVM backend is moving in a direction to support more and more of these)

x86 xmm registers seem to support multiple sizes for a single register class, so maybe amdgpu registers can be modeled like them?

It seems like LLVM and Rust are somewhat forgiving with non-matching types?
E.g. I can assign an i32 to a 16-bit register in x86 (compiles fine though obviously the upper 16-bit of y are garbage afterwards):
```rust
    let x: i32 = 5;
    let y: i32;
    unsafe {
        asm!("mov cx, ax", lateout("cx") y, in("ax") x);
    }
```

Similarly, assigning an i64 to a 32-bit amdgpu sgpr seems to compile fine as well
```rust
check_reg!(s0_i642 i64 "s0" "s_mov_b32");
```

Flakebi commented:
Thanks for the review!
I implemented `overlapping_regs`, it should return about 24 kiB for very large registers, that still seems reasonable.
I realize that the vector type PR overtook this one, so vector types in inline asm is in a half-baked state. Most of the implementation should be generic enough to support them, though the possible types are not exposed and there are no tests yet.
Let me know if you prefer adding vector type support in a separate PR after this one.

Flakebi commented:
Register classes are now split by size and vector types exist now, including tests.

I’m not completely convinced that splitting the register classes by size is the best thing to do.
I implemented it anyway (in a separate commit for now), please let me know if that’s what you had in mind.

> This is not the case for amdgpu since v0 and v[0:1] are very distinct, for example the latter overlaps v1 but the former doesn't.

The overlapping check is defined by register (not by class) and this already works correctly without splitting the classes.
Size checks for fixed registers (e.g. "v0") didn’t wok correctly (similar how they don’t work for x86). Now, trying to assign an i64 to a "v0" gives an error.

apiraino commented:
in case it's relevant, pinging our new GPU-target group

@rustbot ping gpu-target

RalfJung commented:
Uh, what? I am not a GPU expert and never agreed to be in this ping group.^^ I'm afraid I don't have capacity to put GPUs on my plate as well.

Amanieu commented on compiler/rustc_target/src/asm/amdgpu.rs:
```diff
@@ -0,0 +1,486 @@
+        vgpr352,
+        vgpr384,
+        vgpr512,
+        vgpr1024,
```
Considering the sheer variety of register sizes, I think that this should be represented like this instead:
```rust
enum AmdgpuInlineAsmRegClass {
    Sgpr(u16),
    Vgpr(u16),
}
```

Then you would manually implement the relevant methods instead of using the `def_reg_class` macro.

You will need to add validation during `parse` to ensure that the register size is valid: it must be a multiple of 32, with an exception for `{sv}gpr16` which represents half of a normal register.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149793/ef70767064ab87b0a41400f69e1dc0b55c8d5284..246bc6db8ce666646b3c019f927cd0cae60c891a)*

Flakebi replied:
I implemented this in the last commit, it saves 20 lines (yay).
Instead of checking for multiple of 32 I re-used the function that tries converting a class to a symbol. That should be quick to run and is more accurate.

LegNeato commented:
> The fourth register class is AGPRs, which only exist on some hardware versions (not the consumer ones) and they have restricted ways to write and read from them, which makes it hard to write a Rust variable into them. They could be used inside assembly blocks, but I didn’t add them as Rust register class.

AGPRs are critical for datacenter ML workloads, but this seems like the right call to me at least (not as familiar with amd stuff).

Amanieu commented on compiler/rustc_target/src/asm/amdgpu.rs:
```diff
@@ -0,0 +1,496 @@
+    ) -> &'static [(InlineAsmType, Option<Symbol>)] {
+        match self {
+            Self::Vgpr(16) => types! { _: I16, F16; },
+            Self::Sgpr(32) | Self::Vgpr(32) => types! { _: I16, I32, F16, F32,
```
```suggestion
            Self::Sgpr(32) | Self::Vgpr(32) => types! { _: I32, F32,
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149793/7af3402cda75aaead39f72516fd6cbb2f3ee0dbd..4e478e4efd9c0e33a38d08f737fda29a2ac0f804)*

Amanieu commented on compiler/rustc_target/src/asm/amdgpu.rs:
```diff
@@ -0,0 +1,496 @@
+                VecI16(512 / 16), VecI32(512 / 32),
+                VecF16(512 / 16), VecF32(512 / 32);
+            },
+            Self::Vgpr(1024) => types! { _: VecF32(1024 / 32); },
```
Is there a reason for all the inconsistency here? Some sizes support i16/f16 vectors, others support i64/f64 vectors, but I don't see any pattern.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149793/7af3402cda75aaead39f72516fd6cbb2f3ee0dbd..4e478e4efd9c0e33a38d08f737fda29a2ac0f804)*

Flakebi replied:
The supported types are what the llvm backend supports (at the time of writing this code).
I don’t think there’s much of a system there except “there is an instruction that wants this type”.
At the beginning, I added all possible types and then removed those that failed to compile when writing tests for them.

Flakebi replied:
I guess (with the adjacent goal of computing this dynamically), it should be ok to restrict everything to “common” types and remove types that are technically supported but probably not that common.
I will give it a try in the next days. Thanks for your review!

Amanieu commented on compiler/rustc_target/src/asm/amdgpu.rs:
```diff
@@ -0,0 +1,496 @@
+        None
+    }
+
+    pub fn supported_types(
```
It may be better to just rework this function to return a `Cow` and compute the data dynamically.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149793/7af3402cda75aaead39f72516fd6cbb2f3ee0dbd..4e478e4efd9c0e33a38d08f737fda29a2ac0f804)*

Flakebi commented:
`supported_types` is now computed dynamically, returning a `Cow`.
64-bit vector type support is removed in the same step, as that is rather irregular and probably not used much.

I kept support for 16-bit values in 32-bit registers as that is quite commonly used (proper 16-bit register support is only available on newer hardware).

All new changes are in the last commit 581b55a.

Amanieu commented:
@bors r+

JonathanBrouwer commented:
@bors r-
https://triage.rust-lang.org/gha-logs/rust-lang/rust/76430928845

Flakebi committed:
Add inline asm support for amdgpu

Add support for inline assembly for the amdgpu backend (the
amdgcn-amd-amdhsa target).
Add register classes for `vgpr` (vector general purpose register) and
`sgpr` (scalar general purpose register).
The LLVM backend supports two more classes, `reg`, which is either VGPR
or SGPR, up to the compiler to decide. As instructions often rely on a
register being either a VGPR or SGPR for the assembly to be valid, reg
doesn’t seem that useful (I struggled to write correct tests for it), so
I didn’t end up adding it.
The fourth register class is AGPRs, which only exist on some hardware
versions (not the consumer ones) and they have restricted ways to write
and read from them, which makes it hard to write a Rust variable into
them. They could be used inside assembly blocks, but I didn’t add them
as Rust register class.

There are a few change affecting general inline assembly code, that is
`InlineAsmReg::name()` now returns a `Cow` instead of a `&'static str`.
Because amdgpu has many registers, 256 VGPRs plus combinations of 2 or 4
VGPRs, and I didn’t want to list hundreds of static strings, the amdgpu
reg stores the register number(s) and a non-static String is generated
at runtime for the register name.
Similar for register classes and supported_types.

Vectors of 64-bit types are supported by the LLVM backend, but omitted
here to make the code simpler. There is currently no systematic support
in LLVM of which vectors of 64-bit types are supported. Also, they are
likely seldomly unused, vectors of 16- and 32-bit types are important.

Flakebi commented:
Building with debug assertions cought an underflowing subtraction.
I fixed that (which detected overlapping registers in one test) and used the chance to squash all commits and update the description.
The tests pass now for me locally with debug asserts enabled (and I went through the other math for amdgpu asm, it looked fine to me).

Diff:
```diff
diff --git a/compiler/rustc_target/src/asm/amdgpu.rs b/compiler/rustc_target/src/asm/amdgpu.rs
index 191cf6baa26..cc6f612cdb8 100644
--- a/compiler/rustc_target/src/asm/amdgpu.rs
+++ b/compiler/rustc_target/src/asm/amdgpu.rs
@@ -478,7 +478,8 @@ pub fn overlapping_regs(self, mut cb: impl FnMut(AmdgpuInlineAsmReg)) {
             | AmdgpuRegStart::Full(start)) = self.range;

             let size_range = size - 1;
-            for overlap_start in (start - size_range)..=(start + self.class.bytes().div_ceil(4) - 1)
+            for overlap_start in
+                start.saturating_sub(size_range)..=(start + self.class.bytes().div_ceil(4) - 1)
             {
                 let class = match self.class {
                     AmdgpuInlineAsmRegClass::Sgpr(_) => AmdgpuInlineAsmRegClass::Sgpr(size * 32),
diff --git a/tests/assembly-llvm/asm/amdgpu-types.rs b/tests/assembly-llvm/asm/amdgpu-types.rs
index fe8ae88ee83..c68cbe4fd96 100644
--- a/tests/assembly-llvm/asm/amdgpu-types.rs
+++ b/tests/assembly-llvm/asm/amdgpu-types.rs
@@ -223,7 +223,7 @@ macro_rules! check_reg {
 // CHECK: #ASMSTART
 // CHECK: s_load_b128 s{{\[[0-9]+:[0-9]+\]}}, s{{\[[0-9]+:[0-9]+\]}}, s{{[0-9]+}}
 // CHECK: #ASMEND
-check_reg!(s0_i128 i128 "s[0:3]" x: ptr "s[0:1]", y: i32 "s0", "s_load_b128");
+check_reg!(s0_i128 i128 "s[0:3]" x: ptr "s[0:1]", y: i32 "s2", "s_load_b128");

 // CHECK-LABEL: v0_i128:
 // CHECK: #ASMSTART
```

Flakebi commented:
(forgot to mark as ready)
@rustbot ready

Amanieu commented:
@bors r+

Amanieu commented:
@bors delegate+

r=me once conflicts are resolved

Flakebi commented:
@bors r=Amanieu
